### PR TITLE
Addition of Noisy DQN to QLearning

### PR DESCRIPTION
**Pull request by nishantkr18**

### CODE REVIEW

File: src/mlpack/methods/ann/layer/noisylinear_impl.hpp [hunk 1]
```diff
@@ -123,7 +123,7 @@ void NoisyLinear<InputDataType, OutputDataType>::Gradient(
   gradient.rows(weight.n_elem + bias.n_elem, gradient.n_elem - bias.n_elem - 1)
       = arma::vectorise(weightGrad % weightEpsilon);
   gradient.rows(gradient.n_elem - bias.n_elem, gradient.n_elem - 1)
-      = arma::sum(error % biasEpsilon, 1);
+      = arma::sum(error, 1) % biasEpsilon;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/noisylinear_impl.hpp:126]: Nice catch!

**Comment by zoq** [comment]
Hm, not sure why the LaTex Documentation check failed, have to take a closer look into that, but besides that is this ready for a review?

**Comment by nishantkr18** [comment]
> Hm, not sure why the LaTex Documentation check failed, have to take a closer look into that, but besides that is this ready for a review?

Yeah, it fails a lot of times. 
And the Noisy DQN is ready, so you can review that part. Apart from that, I've already started adding Categorical DQN as well. So shall I open a separate PR for that, or continue here?

**Comment by zoq** [comment]
Let's open a new PR for the Categorical DQN, makes it easier to merge to merge this one and review the next PR as well.

**Comment by mrityunjay-tripathi** [comment]
In case that helps, the reason for Latex Documentation failure in the logs ---

Generating docs for compound mlpack::neighbo/home/jenkins-mlpack/workspace/pull-requests doxygen check/src/mlpack/methods/reinforcement_learning/q_networks/**categorical_dqn.hpp:48**: 
error: **The following parameters** of mlpack::rl::CategoricalDQN::CategoricalDQN(const int inputDim, const int h1, const int h2, const int outputDim, const bool isNoisy=false, const size_t atomSize=51) 
**are not documented**:
  **parameter 'isNoisy'
  parameter 'atomSize'** (warning treated as error, aborting now)

**Comment by nishantkr18** [comment]
@mrityunjay-tripathi Thanks for the help :)
@zoq I'll revert the changes right away :+1:

**Comment by zoq** [comment]
> Yeah, it fails a lot of times.

Hm, might be a problem with the somewhat small network, the code looks good to me.

**Comment by zoq** [comment]
We should run the test in a loop to make sure the success rate is reasonable - https://github.com/mlpack/mlpack/issues/922

File: src/mlpack/methods/reinforcement_learning/policy/greedy_policy.hpp [hunk 2]
```diff
@@ -59,14 +59,17 @@ class GreedyPolicy
    *
    * @param actionValue Values for each action.
    * @param deterministic Always select the action greedily.
+   * @param isNoisy Specifies whether the network used is noisy.
    * @return Sampled action.
    */
-  ActionType Sample(const arma::colvec& actionValue, bool deterministic = false)
+  ActionType Sample(const arma::colvec& actionValue,
+      bool deterministic = false,
```
Comment (favre49) [inline src/mlpack/methods/reinforcement_learning/policy/greedy_policy.hpp:66]: Let's indent this to align with the other parameters

File: src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp [hunk 3]
```diff
@@ -96,6 +112,18 @@ class SimpleDQN
     network.ResetParameters();
   }
 
+  /**
+   * Resets noise of the network, is the network is of type noisy.
+   */
+  void ResetNoise()
+  {
+    for(size_t i = 0; i < noisyLayerIndex.size(); i++)
```
Comment (favre49) [inline src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp:120]:
```suggestion
    for (size_t i = 0; i < noisyLayerIndex.size(); i++)
```

**Comment by nishantkr18** [comment]
> We should run the test in a loop to make sure the success rate is reasonable - #922

@zoq So I should run the test in a loop locally right? To ensure it passes all the time?

File: src/mlpack/tests/q_learning_test.cpp [hunk 4]
```diff
@@ -507,55 +320,82 @@ BOOST_AUTO_TEST_CASE(CartPoleWithDuelingDQNPrioritizedReplay)
...
+{
+  // It isn't guaranteed that the network will converge in the specified number
+  // of iterations using random weights.
+  bool converged = false;
+  for (size_t trial = 0; trial < 3; ++trial)
   {
-    double episodeReturn = agent.Episode();
-    averageReturn(episodeReturn);
-    episodes += 1;
+    Log::Debug << "Trail number: " << trial << std::endl;
```
Comment (favre49) [inline src/mlpack/tests/q_learning_test.cpp:343]:
```suggestion
    Log::Debug << "Trial number: " << trial << std::endl;
```

File: src/mlpack/tests/q_learning_test.cpp [hunk 5]
```diff
@@ -507,55 +320,82 @@ BOOST_AUTO_TEST_CASE(CartPoleWithDuelingDQNPrioritizedReplay)
...
 
+//! Test Dueling-Double-Noisy DQN in Cart Pole task.
+BOOST_AUTO_TEST_CASE(CartPoleWithDuelingDoubleNoisyDQN)
+{
+  // It isn't guaranteed that the network will converge in the specified number
+  // of iterations using random weights.
+  bool converged = false;
+  for (size_t trial = 0; trial < 5; ++trial)
+  {
+    Log::Debug << "Trail number: " << trial << std::endl;
```
Comment (favre49) [inline src/mlpack/tests/q_learning_test.cpp:376]:
```suggestion
    Log::Debug << "Trial number: " << trial << std::endl;
```

**Comment by zoq** [comment]
> @zoq So I should run the test in a loop locally right? To ensure it passes all the time?

Yes, ideally all the time, but if it fails once or twice over 100 runs that's fine as well.

**Comment by nishantkr18** [comment]
I'm not sure why the appveyor test keeps failing :thinking: 
Also, on my local machine, the test is able to pass in all the times i tested. Besides, the no of trials are 3 and 5 for simpleNoisy and DuelingNoisy respectively, so I think it should be sufficient? What do u say?

**Comment by kartikdutt18** [comment]
> I'm not sure why the appveyor test keeps failing 

I think that's a timeout error and can be ignored, I think. [Relevant message](https://ci.appveyor.com/project/mlpack/mlpack/builds/33458374/job/mmu5c7a2nx75w0rk/messages)

**Comment by zoq** [comment]
> I'm not sure why the appveyor test keeps failing 🤔
> Also, on my local machine, the test is able to pass in all the times i tested. Besides, the no of trials are 3 and 5 for simpleNoisy and DuelingNoisy respectively, so I think it should be sufficient? What do u say?

No worries about the appveyor issue, as @kartikdutt18 said it's a timeout issue.

About the test, do you think it makes sense to switch to the model architecture you used in the examples repo? Might improve the test pass/fail rate?

**Comment by nishantkr18** [comment]
> About the test, do you think it makes sense to switch to the model architecture you used in the examples repo? Might improve the test pass/fail rate?

If I'm not wrong, in the examples repo, we train the agent for `n` no of episodes, and then look at the performance. And here, we train for `n` no of episodes, but stop as soon as we receive a satisfactory score (i.e. if averageReturn > `thresholdReward`(35 for CartPole)). So dont you think the current one should pass more tests. Because in examplesRepo , we assume that after n episodes, the agent has to perform well; whereas here, during training for n episodes, at anytime if the agent performs well, its enough for us.. 

BTW, right now we use running_stat for storing rewards of each episode. But that way, we consider the agent's initial rewards as well(when it was learning, so was performing badly). Lets say an agent performs very badly(lets say reward = `10`) initially for 200 episodes, but then it starts performing well after that, so for the average return to come close to the threshold(`35` for CartPole), it will take a lot of continuously high rewards. To solve this, why dont we use something like a running_stat with fixed size(with a FIFO pattern) for storing the consecutive rewards? Ideally, this will give a better measure of our agent's performance by ignoring its performance history and taking into account only the last lets say `50` episodes. Here we can set this `50` to an even higher number. This way, we can be sure that the test passes, if the agent converges, not matter how many episodes it takes(but obviously we'll have a max episode number after which the test fails). What do u say?

Also, could u suggest a fixed size FIFO data structure? Maybe we'd need to use `std::vector`? 
Sorry for the long message, I hope I make sense :sweat_smile:

**Comment by zoq** [comment]
> If I'm not wrong, in the examples repo, we train the agent for `n` no of episodes, and then look at the performance. And here, we train for `n` no of episodes, but stop as soon as we receive a satisfactory score (i.e. if averageReturn > `thresholdReward`(35 for CartPole)). So dont you think the current one should pass more tests. Because in examplesRepo , we assume that after n episodes, the agent has to perform well; whereas here, during training for n episodes, at anytime if the agent performs well, its enough for us..

You are right, my intention was to see if the model architecture performs similarly to the one we use in the tests. If it does I think it would be slightly faster, looking at the results in sounded like it often fails in the first and second run.

> BTW, right now we use running_stat for storing rewards of each episode. But that way, we consider the agent's initial rewards as well(when it was learning, so was performing badly). Lets say an agent performs very badly(lets say reward = `10`) initially for 200 episodes, but then it starts performing well after that, so for the average return to come close to the threshold(`35` for CartPole), it will take a lot of continuously high rewards. To solve this, why dont we use something like a running_stat with fixed size(with a FIFO pattern) for storing the consecutive rewards? Ideally, this will give a better measure of our agent's performance by ignoring its performance history and taking into account only the last lets say `50` episodes. Here we can set this `50` to an even higher number. This way, we can be sure that the test passes, if the agent converges, not matter how many episodes it takes(but obviously we'll have a max episode number after which the test fails). What do u say?

I see what you mean, the advantage of the current method is that we make sure the model improves over time and stays at a high level, but I can see the downside for this especially for testing. Unfortunately you can't set the maximum number of examples for the running_stat module you can only reset it, so we would have to reimplement some of the functionality.

boost circular buffer would be an option, but there are some thoughts about moving away from boost so it might not be the best option, `std::vector` should work as well, we just need a list that allows us to push a new element to the end and remove the first one.

**Comment by nishantkr18** [comment]
`'accumulate': is not a member of 'std'` for windows builds. Should I go with the loop method to find the mean?
``` c++
double avg = 0;
for (size_t i = 0; i < v.size(); i++)
{
  avg += v[i];
}
avg /= v.size();
```
?

**Comment by favre49** [comment]
Shouldn't `accumulate` work if we add `#include <numeric>` ? If not, we could go with the loop choice

File: src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp [hunk 6]
```diff
@@ -211,6 +211,12 @@ double QLearning<
       gradients);
   #endif
 
+  if(config.NoisyQLearning() == true)
```
Comment (favre49) [inline src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp:214]:
```suggestion
  if (config.NoisyQLearning() == true)
```

**Comment by zoq** [comment]
Let's wait for @birm if he has any comments before we merge this in.

**Pull request merged**